### PR TITLE
Support paw wishes

### PR DESCRIPTION
**Pull request by Pantocyclus**

### CODE REVIEW

File: src/potions.ts [hunk 1]
```diff
@@ -79,6 +82,27 @@ for (const effectGroup of mutuallyExclusiveList) {
...
+  if (!validWishString.has(effect)) {
+    const allOtherNoHookahEffectNames = allNoHookahEffects
+      .filter((e) => e !== effect)
+      .map((e) => e.name.toLowerCase());
+    const validSubstrings = Array.from(
+      effect.name.toLowerCase().match(RegExp(/[\w ]+/g)) ?? []
+    ).filter((s) => !allOtherNoHookahEffectNames.some((name) => name.includes(s)));
+    validWishString.set(
+      effect,
+      (validSubstrings.length > 0 ? validSubstrings.pop() ?? "" : "").trim()
```
Comment (horrible-little-slime) [inline src/potions.ts:97]:
```suggestion
      (validSubstrings[0]?.trim() ?? "")
```

File: src/potions.ts [hunk 2]
```diff
@@ -460,18 +484,64 @@ const rufusPotion = new Potion($item`closed-circuit pay phone`, {
...
+export const wishPotions = validWishes.map(
+  (effect) =>
+    new Potion($item`pocket wish`, {
+      effect,
+      canDouble: false,
+      duration: 20,
+      use: (quantity: number) =>
+        new Array(quantity).fill(0).every(() => {
+          const s = getValidWishString(effect);
+          if (s === "") return;
```
Comment (horrible-little-slime) [inline src/potions.ts:496]:
```suggestion
          if (!s) return;
```

File: src/potions.ts [hunk 3]
```diff
@@ -460,18 +484,64 @@ const rufusPotion = new Potion($item`closed-circuit pay phone`, {
...
+          cliExecute(`genie effect ${s}`);
+        }),
+    })
```
Comment (horrible-little-slime) [inline src/potions.ts:499]: This is now long enough that we can't justify using Array.fill.every, and should instead use a `for` loop in a code block

File: src/potions.ts [hunk 4]
```diff
@@ -460,18 +484,64 @@ const rufusPotion = new Potion($item`closed-circuit pay phone`, {
...
         canDouble: false,
-        duration: 20,
-        use: (quantity: number) =>
-          new Array(quantity).fill(0).every(() => cliExecute(`genie effect ${effect}`)),
+        price: () =>
+          get("_pawWishes", 0) >= 5 ||
+          // eslint-disable-next-line libram/verify-constants
+          !have($item`cursed monkey's paw`) ||
+          failedWishes.includes(effect)
+            ? Infinity
```
Comment (horrible-little-slime) [inline src/potions.ts:514]: I love using `Infinity`, but keep in mind that if we ever `sort` the array it'll cause issues (the `reduce` used in `maxBy` is fine, though)

File: src/potions.ts [hunk 5]
```diff
@@ -544,6 +614,30 @@ export function potionSetup(embezzlersOnly: boolean): void {
...
+    if (
+      !pawPotions
+        .filter((potion) => numericModifier(potion.effect(), "Meat Drop") >= 100)
+        .sort((a, b) => {
+          const aValue = a
+            .value(embezzlers)
+            .sort((l, r) => (r.quantity > 0 ? r.value : 0) - (l.quantity > 0 ? l.value : 0));
+          const bValue = b
+            .value(embezzlers)
+            .sort((l, r) => (r.quantity > 0 ? r.value : 0) - (l.quantity > 0 ? l.value : 0));
```
Comment (horrible-little-slime) [inline src/potions.ts:630]: this is hard to follow and probably should just use a `maxBy` given that you're ultimately checking their highest value

File: src/potions.ts [hunk 6]
```diff
@@ -544,6 +614,30 @@ export function potionSetup(embezzlersOnly: boolean): void {
...
+          return (
+            (bValue.length > 0 ? bValue[0].value : 0) - (aValue.length > 0 ? aValue[0].value : 0)
+          );
+        })
+        .some((potion) => potion.use(1))
```
Comment (horrible-little-slime) [inline src/potions.ts:635]: This also does not seem like a great structure to end up in

File: src/potions.ts (hunk 2 above)
Comment (horrible-little-slime) [reply src/potions.ts:496]: honestly, given that we always check against "is an empty string", using `null` or `undefined` instead of `""` is going to be easier to follow imo

**Review by horrible-little-slime — CHANGES_REQUESTED** [review]
I think @pstalcup will have more to say in the coming days, and given that the tracking PR is now open in mafia (thanks Veracity!), we should just wait for that and update our mafia version.

I'll cut a new eslint-plugin-libram now

File: src/potions.ts (hunk 4 above)
Comment (Pantocyclus) [reply src/potions.ts:511]: I think it's fine. We should never want to use a potion that has a cost of Infinity

File: src/potions.ts [hunk 7]
```diff
@@ -79,6 +82,24 @@ for (const effectGroup of mutuallyExclusiveList) {
...
+    const allOtherNoHookahEffectNames = allNoHookahEffects
+      .filter((e) => e !== effect)
+      .map((e) => e.name.toLowerCase());
+    const validSubstrings = Array.from(
+      effect.name.toLowerCase().match(RegExp(/[\w ]+/g)) ?? []
+    ).filter((s) => !allOtherNoHookahEffectNames.some((name) => name.includes(s)));
+    validWishString.set(effect, validSubstrings[0]?.trim() ?? "");
+  }
+  return validWishString.get(effect) ?? "";
+}
```
Comment (horrible-little-slime) [inline src/potions.ts:98]: use `null` as your dummy value instead of the empty string

File: src/potions.ts [hunk 8]
```diff
@@ -79,6 +82,24 @@ for (const effectGroup of mutuallyExclusiveList) {
...
+const validWishes = allNoHookahEffects.filter((effect) => {
+  return getValidWishString(effect)?.length > 0;
+});
```
Comment (horrible-little-slime) [inline src/potions.ts:101]:
```suggestion
const validWishes = allNoHookahEffects.filter((effect) => getValidWishString(effect));
```

File: src/potions.ts [hunk 9]
```diff
@@ -460,18 +481,64 @@ const rufusPotion = new Potion($item`closed-circuit pay phone`, {
...
+    new Potion($item`pocket wish`, {
+      effect,
+      canDouble: false,
+      duration: 20,
+      use: (quantity: number) => {
+        const s = getValidWishString(effect);
+        if (!s) return false;
+        return new Array(quantity).fill(0).every(() => {
+          cliExecute(`genie effect ${s}`);
+        });
```
Comment (horrible-little-slime) [inline src/potions.ts:495]: Use a `for` loop here

File: src/potions.ts [hunk 10]
```diff
@@ -460,18 +481,64 @@ const rufusPotion = new Potion($item`closed-circuit pay phone`, {
...
+      },
+    })
+);
+
+export const pawPotions = validWishes
   .map(
     (effect) =>
-      new Potion($item`pocket wish`, {
+      // eslint-disable-next-line libram/verify-constants
+      new Potion($item`cursed monkey's paw`, {
```
Comment (horrible-little-slime) [inline src/potions.ts:504]: You can bump the eslint-plugin-libram version for this

File: src/potions.ts [hunk 11]
```diff
@@ -493,6 +560,20 @@ export function doublingPotions(embezzlers: number): Potion[] {
...
 }
 
+export function usePawWishes(singleUseValuation: (potion: Potion) => number): void {
+  while (get("_pawWishes", 0) < 5) {
+    // Sort the paw potions by the profits of a single wish, then use the best one
+    if (
+      !pawPotions
+        .filter((potion) => numericModifier(potion.effect(), "Meat Drop") >= 100)
+        .sort((a, b) => singleUseValuation(b) - singleUseValuation(a))
+        .some((potion) => potion.use(1))
```
Comment (horrible-little-slime) [inline src/potions.ts:570]: Don't put effectful code in this `if` statement

File: src/potions.ts [hunk 12]
```diff
@@ -493,6 +560,20 @@ export function doublingPotions(embezzlers: number): Potion[] {
     .map((pair) => pair.potion);
 }
 
+export function usePawWishes(singleUseValuation: (potion: Potion) => number): void {
+  while (get("_pawWishes", 0) < 5) {
+    // Sort the paw potions by the profits of a single wish, then use the best one
+    if (
+      !pawPotions
+        .filter((potion) => numericModifier(potion.effect(), "Meat Drop") >= 100)
+        .sort((a, b) => singleUseValuation(b) - singleUseValuation(a))
```
Comment (horrible-little-slime) [inline src/potions.ts:569]: This will behave very poorly with `Infinity` in the array

File: src/potions.ts [hunk 13]
```diff
@@ -493,6 +560,20 @@ export function doublingPotions(embezzlers: number): Potion[] {
     .map((pair) => pair.potion);
 }
 
+export function usePawWishes(singleUseValuation: (potion: Potion) => number): void {
+  while (get("_pawWishes", 0) < 5) {
+    // Sort the paw potions by the profits of a single wish, then use the best one
+    if (
+      !pawPotions
+        .filter((potion) => numericModifier(potion.effect(), "Meat Drop") >= 100)
```
Comment (horrible-little-slime) [inline src/potions.ts:568]: You already do this filter above; is this redundant?

File: src/potions.ts [hunk 14]
```diff
@@ -493,6 +560,20 @@ export function doublingPotions(embezzlers: number): Potion[] {
...
+        .sort((a, b) => singleUseValuation(b) - singleUseValuation(a))
+        .some((potion) => potion.use(1))
+    ) {
+      break;
```
Comment (horrible-little-slime) [inline src/potions.ts:572]: use `return` instead of `break` if there's nothing after the `while` loop

**Review by horrible-little-slime — CHANGES_REQUESTED** [review]

**Comment by horrible-little-slime** [comment]
This also needs a kolmafia version bump

File: src/potions.ts (hunk 11 above)
Comment (horrible-little-slime) [reply src/potions.ts:570]: Or, rather, not this far buried in the `if` statement--obviously you need to capture and evaluate the value somehow

File: src/potions.ts [hunk 15]
```diff
@@ -79,6 +81,26 @@ for (const effectGroup of mutuallyExclusiveList) {
...
+    ).filter((s) => !allOtherNoHookahEffectNames.some((name) => name.includes(s)));
+    validWishString.set(effect, validSubstrings[0]?.trim());
+  }
+  return validWishString.get(effect);
+}
+const validWishes = allNoHookahEffects
+  .filter((effect) => getValidWishString(effect))
+  .filter(
+    (effect) => (globalOptions.prefs.yachtzeechain ? effect !== $effect`Eau d' Clochard` : true) // hardcoded heuristics
+  );
```
Comment (horrible-little-slime) [inline src/potions.ts:102]: These can/should probably be merged

Comment (horrible-little-slime) [reply src/potions.ts:102]:
```suggestion
  .filter((effect) => !!getValidWishString(effect) && (!globalOptions.prefs.yachtzeechain || effect !== $effect`Eau d' Clochard`));
```

you may or may not need the double bang here

File: src/potions.ts [hunk 16]
```diff
@@ -460,18 +482,64 @@ const rufusPotion = new Potion($item`closed-circuit pay phone`, {
...
-export const wishPotions = Effect.all()
-  .filter((effect) => !effect.attributes.includes("nohookah"))
+export const wishPotions = validWishes.map(
+  (effect) =>
+    new Potion($item`pocket wish`, {
+      effect,
+      canDouble: false,
+      duration: 20,
+      use: (quantity: number) => {
+        const s = getValidWishString(effect);
```
Comment (horrible-little-slime) [inline src/potions.ts:492]: you don't need this for genie, afaik; I don't know why I didn't notice this earlier lol

File: src/potions.ts [hunk 17]
```diff
@@ -460,18 +482,64 @@ const rufusPotion = new Potion($item`closed-circuit pay phone`, {
...
+          }
+
+          const s = getValidWishString(effect);
+          if (!s) return false;
+
+          const currentEffectTurns = haveEffect(effect);
+          visitUrl("main.php?action=cmonk&pwd");
+          runChoice(1, `wish=${s}`);
+          visitUrl("main.php");
+          if (haveEffect(effect) <= currentEffectTurns) {
```
Comment (horrible-little-slime) [inline src/potions.ts:534]: imo it would be better practice to regexp the result of runChoice, but that can wait for libram

File: src/potions.ts [hunk 18]
```diff
@@ -79,6 +81,26 @@ for (const effectGroup of mutuallyExclusiveList) {
   }
 }
 
+const allNoHookahEffects = Effect.all().filter((effect) => !effect.attributes.includes("nohookah"));
+const validWishString: Map<Effect, string> = new Map();
+function getValidWishString(effect: Effect): string | undefined {
```
Comment (horrible-little-slime) [inline src/potions.ts:86]: extremely funny thought: do we need to check these substrings against item names as well?

**Review by horrible-little-slime — CHANGES_REQUESTED** [review]
This really feels like the last round of feedback! Great work!

File: src/potions.ts (hunk 16 above)
Comment (horrible-little-slime) [reply src/potions.ts:492]: if you do need special logic for genie strings, it will be different from the special logic needed for monkey strings, because life is hell

@s-k-z knows a lot about this iirc, and may have an open libram PR that I should merge

File: src/potions.ts (hunk 18 above)
Comment (horrible-little-slime) [reply src/potions.ts:86]: update:
we only care about if a substring is an exact match for an available item name. This will probably? never happen.
But!
```
const availableItems = [...new Set(Location.all().filter((l) => canAdventure(l)).map((l) => getMonsters(l)).filter((m) => m.copyable).map((m) => itemDropsArray(m)).flat())];
```
or something like that

Comment (Pantocyclus) [reply src/potions.ts]: I added this in, but we should also keep in mind that this is created at the instance of invoking garbo. We may gain (one-day tickets, transponder etc) and/or lose (content unlocker effects running out) access to locations as garbo runs, so this will then be a slightly incorrect list. 

(But I don't think we should be updating this every turn, so this is the best we can do)

**Review by horrible-little-slime — APPROVED** [review]

File: src/yachtzee/buffs.ts [hunk 19]
```diff
@@ -67,6 +68,13 @@ export function yachtzeePotionProfits(potion: Potion, yachtzeeTurns: number): nu
   return yachtzeeValue + embezzlerValue + barfValue - potion.price(true);
 }
 
+const doublingValue = (potion: Potion, yachtzeeTurns: number) =>
+  Math.min(
+    potion.price(false),
+    yachtzeePotionProfits(potion.doubleDuration(), yachtzeeTurns) -
+      Math.max(0, yachtzeePotionProfits(potion, yachtzeeTurns))
+  );
+
```
Comment (horrible-little-slime) [inline src/yachtzee/buffs.ts:77]:
```suggestion
const doublingValue = (yachtzeeTurns: number) => (potion: Potion) =>
  Math.min(
    potion.price(false),
    yachtzeePotionProfits(potion.doubleDuration(), yachtzeeTurns) -
      Math.max(0, yachtzeePotionProfits(potion, yachtzeeTurns))
  );
```

Would this look better as a curried function, letting us do `maxBy(doublingPotions, doublingValue(yachtzeeTurns))`?
This isn't actively a suggestion, just a pitch

Comment (Pantocyclus) [reply src/yachtzee/buffs.ts:77]: This is the same as `partial()` in Python, so it makes sense. I just don't know whether it's a common stylistic thing in JS/TS; if so I can change it

**Review by horrible-little-slime — DISMISSED** [review]
